<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_121) on Fri Feb 10 23:15:27 CET 2017 -->
<title>AnalysedSequence</title>
<meta name="date" content="2017-02-10">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
<script type="text/javascript" src="../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="AnalysedSequence";
        }
    }
    catch(err) {
    }
//-->
var methods = {"i0":10,"i1":10,"i2":10,"i3":10,"i4":10,"i5":10,"i6":10,"i7":10,"i8":10,"i9":10,"i10":10,"i11":10,"i12":10,"i13":10,"i14":10,"i15":10,"i16":10,"i17":10,"i18":10,"i19":10,"i20":10,"i21":10,"i22":10,"i23":10,"i24":10,"i25":10,"i26":10};
var tabs = {65535:["t0","All Methods"],2:["t2","Instance Methods"],8:["t4","Concrete Methods"]};
var altColor = "altColor";
var rowColor = "rowColor";
var tableTab = "tableTab";
var activeTableTab = "activeTableTab";
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/AnalysedSequence.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev&nbsp;Class</li>
<li><a href="../analysis/Gene.html" title="class in analysis"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?analysis/AnalysedSequence.html" target="_top">Frames</a></li>
<li><a href="AnalysedSequence.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">analysis</div>
<h2 title="Class AnalysedSequence" class="title">Class AnalysedSequence</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li>java.lang.Object</li>
<li>
<ul class="inheritance">
<li><a href="../analysis/Sequence.html" title="class in analysis">analysis.Sequence</a></li>
<li>
<ul class="inheritance">
<li>analysis.AnalysedSequence</li>
</ul>
</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<hr>
<br>
<pre>public class <span class="typeNameLabel">AnalysedSequence</span>
extends <a href="../analysis/Sequence.html" title="class in analysis">Sequence</a></pre>
<div class="block">Models a sequence under analysis (i.e. obtained from an AB1 file), which may have mutations. The
 Sequence class defines it's basic behavior.</div>
<dl>
<dt><span class="simpleTagLabel">Author:</span></dt>
<dd>Ben Kohr, Jannis Blueml</dd>
</dl>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#AnalysedSequence-java.lang.String-java.lang.String-java.lang.String-int:A-">AnalysedSequence</a></span>(java.lang.String&nbsp;sequence,
                java.lang.String&nbsp;researcher,
                java.lang.String&nbsp;fileName,
                int[]&nbsp;qualities)</code>
<div class="block">Constructor calling the super constructor (which sets all given attributes).</div>
</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method.summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span id="t0" class="activeTableTab"><span>All Methods</span><span class="tabEnd">&nbsp;</span></span><span id="t2" class="tableTab"><span><a href="javascript:show(2);">Instance Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t4" class="tableTab"><span><a href="javascript:show(8);">Concrete Methods</a></span><span class="tabEnd">&nbsp;</span></span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr id="i0" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#addMutation-java.lang.String-">addMutation</a></span>(java.lang.String&nbsp;mutation)</code>
<div class="block">Add a discovered, String-encoded mutation to the list of already discovered mutations.</div>
</td>
</tr>
<tr id="i1" class="rowColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#getAvgQuality--">getAvgQuality</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i2" class="altColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#getComments--">getComments</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i3" class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#getFileName--">getFileName</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i4" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#getHisTagPosition--">getHisTagPosition</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i5" class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#getLeftVector--">getLeftVector</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i6" class="altColor">
<td class="colFirst"><code>java.util.LinkedList&lt;java.lang.String&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#getMutations--">getMutations</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i7" class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#getOffset--">getOffset</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i8" class="altColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#getPrimer--">getPrimer</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i9" class="rowColor">
<td class="colFirst"><code>int[]</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#getQuality--">getQuality</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i10" class="altColor">
<td class="colFirst"><code><a href="../analysis/Gene.html" title="class in analysis">Gene</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#getReferencedGene--">getReferencedGene</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i11" class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#getRightVector--">getRightVector</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i12" class="altColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#getTrimPercentage--">getTrimPercentage</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i13" class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#isManuallyChecked--">isManuallyChecked</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i14" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#length--">length</a></span>()</code>
<div class="block">Returns the length of the sequence (the number of nucleotides in it).</div>
</td>
</tr>
<tr id="i15" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#reverseQuality--">reverseQuality</a></span>()</code>
<div class="block">This method reverses the Qualityarray and set it new.</div>
</td>
</tr>
<tr id="i16" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#setComments-java.lang.String-">setComments</a></span>(java.lang.String&nbsp;comments)</code>&nbsp;</td>
</tr>
<tr id="i17" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#setHisTagPosition-int-">setHisTagPosition</a></span>(int&nbsp;hisTagPosition)</code>&nbsp;</td>
</tr>
<tr id="i18" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#setLeftVector-java.lang.String-">setLeftVector</a></span>(java.lang.String&nbsp;vector)</code>&nbsp;</td>
</tr>
<tr id="i19" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#setManuallyChecked-boolean-">setManuallyChecked</a></span>(boolean&nbsp;manuallyChecked)</code>&nbsp;</td>
</tr>
<tr id="i20" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#setOffset-int-">setOffset</a></span>(int&nbsp;offset)</code>&nbsp;</td>
</tr>
<tr id="i21" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#setPrimer-java.lang.String-">setPrimer</a></span>(java.lang.String&nbsp;primer)</code>&nbsp;</td>
</tr>
<tr id="i22" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#setReferencedGene-analysis.Gene-">setReferencedGene</a></span>(<a href="../analysis/Gene.html" title="class in analysis">Gene</a>&nbsp;gene)</code>&nbsp;</td>
</tr>
<tr id="i23" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#setRightVector-java.lang.String-">setRightVector</a></span>(java.lang.String&nbsp;vector)</code>&nbsp;</td>
</tr>
<tr id="i24" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#setTrimPercentage-double-">setTrimPercentage</a></span>(double&nbsp;trimPercentage)</code>&nbsp;</td>
</tr>
<tr id="i25" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#trimQualityArray-int-int-">trimQualityArray</a></span>(int&nbsp;startIndex,
                int&nbsp;endIndex)</code>
<div class="block">This method trims a quality array, i.e.</div>
</td>
</tr>
<tr id="i26" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/AnalysedSequence.html#trimSequence-int-int-">trimSequence</a></span>(int&nbsp;startIndex,
            int&nbsp;endIndex)</code>
<div class="block">This method trims a sequence, i.e.</div>
</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.analysis.Sequence">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;analysis.<a href="../analysis/Sequence.html" title="class in analysis">Sequence</a></h3>
<code><a href="../analysis/Sequence.html#getAddingDate--">getAddingDate</a>, <a href="../analysis/Sequence.html#getComplementarySequence--">getComplementarySequence</a>, <a href="../analysis/Sequence.html#getComplementarySequence-java.lang.String-">getComplementarySequence</a>, <a href="../analysis/Sequence.html#getResearcher--">getResearcher</a>, <a href="../analysis/Sequence.html#getReversedSequence--">getReversedSequence</a>, <a href="../analysis/Sequence.html#getSequence--">getSequence</a>, <a href="../analysis/Sequence.html#setAddingDate-java.lang.String-">setAddingDate</a>, <a href="../analysis/Sequence.html#setResearcher-java.lang.String-">setResearcher</a>, <a href="../analysis/Sequence.html#setSequence-java.lang.String-">setSequence</a>, <a href="../analysis/Sequence.html#toString--">toString</a></code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.Object</h3>
<code>equals, getClass, hashCode, notify, notifyAll, wait, wait, wait</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="AnalysedSequence-java.lang.String-java.lang.String-java.lang.String-int:A-">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>AnalysedSequence</h4>
<pre>public&nbsp;AnalysedSequence(java.lang.String&nbsp;sequence,
                        java.lang.String&nbsp;researcher,
                        java.lang.String&nbsp;fileName,
                        int[]&nbsp;qualities)</pre>
<div class="block">Constructor calling the super constructor (which sets all given attributes).</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>sequence</code> - the nucleotide sequence as a String</dd>
<dd><code>researcher</code> - the researcher's name</dd>
<dd><code>fileName</code> - the name of the file this sequence was obtained from</dd>
<dd><code>qualities</code> - the int-array of quality measurements</dd>
<dd><code>average</code> - the average quality</dd>
</dl>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method.detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="addMutation-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>addMutation</h4>
<pre>public&nbsp;void&nbsp;addMutation(java.lang.String&nbsp;mutation)</pre>
<div class="block">Add a discovered, String-encoded mutation to the list of already discovered mutations.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>mutation</code> - A discovered mutation</dd>
</dl>
</li>
</ul>
<a name="getAvgQuality--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getAvgQuality</h4>
<pre>public&nbsp;double&nbsp;getAvgQuality()</pre>
</li>
</ul>
<a name="getComments--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getComments</h4>
<pre>public&nbsp;java.lang.String&nbsp;getComments()</pre>
</li>
</ul>
<a name="getFileName--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getFileName</h4>
<pre>public&nbsp;java.lang.String&nbsp;getFileName()</pre>
</li>
</ul>
<a name="getTrimPercentage--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getTrimPercentage</h4>
<pre>public&nbsp;double&nbsp;getTrimPercentage()</pre>
</li>
</ul>
<a name="getLeftVector--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getLeftVector</h4>
<pre>public&nbsp;java.lang.String&nbsp;getLeftVector()</pre>
</li>
</ul>
<a name="getMutations--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getMutations</h4>
<pre>public&nbsp;java.util.LinkedList&lt;java.lang.String&gt;&nbsp;getMutations()</pre>
</li>
</ul>
<a name="getOffset--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getOffset</h4>
<pre>public&nbsp;int&nbsp;getOffset()</pre>
</li>
</ul>
<a name="getPrimer--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getPrimer</h4>
<pre>public&nbsp;java.lang.String&nbsp;getPrimer()</pre>
</li>
</ul>
<a name="getQuality--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getQuality</h4>
<pre>public&nbsp;int[]&nbsp;getQuality()</pre>
</li>
</ul>
<a name="getHisTagPosition--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getHisTagPosition</h4>
<pre>public&nbsp;int&nbsp;getHisTagPosition()</pre>
</li>
</ul>
<a name="getReferencedGene--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getReferencedGene</h4>
<pre>public&nbsp;<a href="../analysis/Gene.html" title="class in analysis">Gene</a>&nbsp;getReferencedGene()</pre>
</li>
</ul>
<a name="getRightVector--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getRightVector</h4>
<pre>public&nbsp;java.lang.String&nbsp;getRightVector()</pre>
</li>
</ul>
<a name="isManuallyChecked--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isManuallyChecked</h4>
<pre>public&nbsp;boolean&nbsp;isManuallyChecked()</pre>
</li>
</ul>
<a name="length--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>length</h4>
<pre>public&nbsp;int&nbsp;length()</pre>
<div class="block">Returns the length of the sequence (the number of nucleotides in it).</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the sequence's length</dd>
</dl>
</li>
</ul>
<a name="reverseQuality--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>reverseQuality</h4>
<pre>public&nbsp;void&nbsp;reverseQuality()</pre>
<div class="block">This method reverses the Qualityarray and set it new.</div>
</li>
</ul>
<a name="setComments-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setComments</h4>
<pre>public&nbsp;void&nbsp;setComments(java.lang.String&nbsp;comments)</pre>
</li>
</ul>
<a name="setLeftVector-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setLeftVector</h4>
<pre>public&nbsp;void&nbsp;setLeftVector(java.lang.String&nbsp;vector)</pre>
</li>
</ul>
<a name="setManuallyChecked-boolean-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setManuallyChecked</h4>
<pre>public&nbsp;void&nbsp;setManuallyChecked(boolean&nbsp;manuallyChecked)</pre>
</li>
</ul>
<a name="setOffset-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setOffset</h4>
<pre>public&nbsp;void&nbsp;setOffset(int&nbsp;offset)</pre>
</li>
</ul>
<a name="setPrimer-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setPrimer</h4>
<pre>public&nbsp;void&nbsp;setPrimer(java.lang.String&nbsp;primer)</pre>
</li>
</ul>
<a name="setReferencedGene-analysis.Gene-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setReferencedGene</h4>
<pre>public&nbsp;void&nbsp;setReferencedGene(<a href="../analysis/Gene.html" title="class in analysis">Gene</a>&nbsp;gene)</pre>
</li>
</ul>
<a name="setRightVector-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setRightVector</h4>
<pre>public&nbsp;void&nbsp;setRightVector(java.lang.String&nbsp;vector)</pre>
</li>
</ul>
<a name="setHisTagPosition-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setHisTagPosition</h4>
<pre>public&nbsp;void&nbsp;setHisTagPosition(int&nbsp;hisTagPosition)</pre>
</li>
</ul>
<a name="setTrimPercentage-double-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setTrimPercentage</h4>
<pre>public&nbsp;void&nbsp;setTrimPercentage(double&nbsp;trimPercentage)</pre>
</li>
</ul>
<a name="trimQualityArray-int-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>trimQualityArray</h4>
<pre>public&nbsp;void&nbsp;trimQualityArray(int&nbsp;startIndex,
                             int&nbsp;endIndex)</pre>
<div class="block">This method trims a quality array, i.e. it cuts out the desired part of the nucleotide sequence
 out of it. It keeps the quality for the start index character, and all following characters
 including the end index character. The rest is discarded. The result is stored within the
 object.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>startIndex</code> - The start of the array to be cut off</dd>
<dd><code>endIndex</code> - The end of the array to be cut off</dd>
</dl>
</li>
</ul>
<a name="trimSequence-int-int-">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>trimSequence</h4>
<pre>public&nbsp;void&nbsp;trimSequence(int&nbsp;startIndex,
                         int&nbsp;endIndex)</pre>
<div class="block">This method trims a sequence, i.e. it cuts out the desired part of the nucleotide sequence. It
 keeps the start index character, and all following characters including the end index
 character. The rest is discarded. The result is stored within the object.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>startIndex</code> - The start of the sequence to be cut off</dd>
<dd><code>endIndex</code> - The end of the sequence to be cut off</dd>
</dl>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/AnalysedSequence.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev&nbsp;Class</li>
<li><a href="../analysis/Gene.html" title="class in analysis"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?analysis/AnalysedSequence.html" target="_top">Frames</a></li>
<li><a href="AnalysedSequence.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
