<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_121) on Fri Feb 10 23:15:27 CET 2017 -->
<title>StringAnalysis</title>
<meta name="date" content="2017-02-10">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
<script type="text/javascript" src="../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="StringAnalysis";
        }
    }
    catch(err) {
    }
//-->
var methods = {"i0":9,"i1":9,"i2":9,"i3":9,"i4":9,"i5":9,"i6":9,"i7":41,"i8":9,"i9":9,"i10":9,"i11":9,"i12":9,"i13":9,"i14":9,"i15":41,"i16":9,"i17":9};
var tabs = {65535:["t0","All Methods"],1:["t1","Static Methods"],8:["t4","Concrete Methods"],32:["t6","Deprecated Methods"]};
var altColor = "altColor";
var rowColor = "rowColor";
var tableTab = "tableTab";
var activeTableTab = "activeTableTab";
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/StringAnalysis.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../analysis/Sequence.html" title="class in analysis"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li>Next&nbsp;Class</li>
</ul>
<ul class="navList">
<li><a href="../index.html?analysis/StringAnalysis.html" target="_top">Frames</a></li>
<li><a href="StringAnalysis.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">analysis</div>
<h2 title="Class StringAnalysis" class="title">Class StringAnalysis</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li>java.lang.Object</li>
<li>
<ul class="inheritance">
<li>analysis.StringAnalysis</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<hr>
<br>
<pre>public class <span class="typeNameLabel">StringAnalysis</span>
extends java.lang.Object</pre>
<div class="block">This class contains the logic of analyzing sequence strings. This class serves for
 MutationAnaysis by providing useful String Matching methods.</div>
<dl>
<dt><span class="simpleTagLabel">Author:</span></dt>
<dd>jannis blueml</dd>
</dl>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- =========== FIELD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="field.summary">
<!--   -->
</a>
<h3>Field Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Field Summary table, listing fields, and an explanation">
<caption><span>Fields</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Field and Description</th>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.util.Map&lt;java.lang.String,java.lang.String&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/StringAnalysis.html#AMINO_ACID_SHORTS">AMINO_ACID_SHORTS</a></span></code>
<div class="block">A Map with all possible RNA and DNA codons with the matched AminoAcid in short form.</div>
</td>
</tr>
</table>
</li>
</ul>
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../analysis/StringAnalysis.html#StringAnalysis--">StringAnalysis</a></span>()</code>&nbsp;</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method.summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span id="t0" class="activeTableTab"><span>All Methods</span><span class="tabEnd">&nbsp;</span></span><span id="t1" class="tableTab"><span><a href="javascript:show(1);">Static Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t4" class="tableTab"><span><a href="javascript:show(8);">Concrete Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t6" class="tableTab"><span><a href="javascript:show(32);">Deprecated Methods</a></span><span class="tabEnd">&nbsp;</span></span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr id="i0" class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/StringAnalysis.html#appentStringToLength-java.lang.String-int-">appentStringToLength</a></span>(java.lang.String&nbsp;input,
                    int&nbsp;length)</code>&nbsp;</td>
</tr>
<tr id="i1" class="rowColor">
<td class="colFirst"><code>static int[][]</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/StringAnalysis.html#calculateLevenshteinMatrix-java.lang.String-java.lang.String-">calculateLevenshteinMatrix</a></span>(java.lang.String&nbsp;horizontal,
                          java.lang.String&nbsp;vertical)</code>
<div class="block">Calculates the Levensthein Matrix of two Strings.</div>
</td>
</tr>
<tr id="i2" class="altColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/StringAnalysis.html#checkComplementAndReverse-analysis.AnalysedSequence-">checkComplementAndReverse</a></span>(<a href="../analysis/AnalysedSequence.html" title="class in analysis">AnalysedSequence</a>&nbsp;toAnalyse)</code>
<div class="block">This method checks if the complementary and/or the reverse gene has a better similarity with
 the mutated sequence the the original gene.</div>
</td>
</tr>
<tr id="i3" class="rowColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/StringAnalysis.html#checkComplementAndReverse-analysis.AnalysedSequence-analysis.Gene-">checkComplementAndReverse</a></span>(<a href="../analysis/AnalysedSequence.html" title="class in analysis">AnalysedSequence</a>&nbsp;toAnalyse,
                         <a href="../analysis/Gene.html" title="class in analysis">Gene</a>&nbsp;gene)</code>
<div class="block">This method is used when the gene is not the referenced gene in the AnalysedSequence Note: This
 method overrides the old referenced gene with the given gene</div>
</td>
</tr>
<tr id="i4" class="altColor">
<td class="colFirst"><code>static double</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/StringAnalysis.html#checkSimilarity-analysis.Sequence-analysis.Sequence-">checkSimilarity</a></span>(<a href="../analysis/Sequence.html" title="class in analysis">Sequence</a>&nbsp;first,
               <a href="../analysis/Sequence.html" title="class in analysis">Sequence</a>&nbsp;second)</code>
<div class="block">Compares to sequences and returns their similarity without finding the exact differences.</div>
</td>
</tr>
<tr id="i5" class="rowColor">
<td class="colFirst"><code>static double</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/StringAnalysis.html#checkSimilarity-java.lang.String-java.lang.String-">checkSimilarity</a></span>(java.lang.String&nbsp;first,
               java.lang.String&nbsp;second)</code>
<div class="block">Compares to String and returns their similarity without finding the exact differences.</div>
</td>
</tr>
<tr id="i6" class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/StringAnalysis.html#codonsToAminoAcids-java.lang.String-">codonsToAminoAcids</a></span>(java.lang.String&nbsp;nucleotides)</code>
<div class="block">Changes the sequence representation from nucleotides to aminoacid (shortform)</div>
</td>
</tr>
<tr id="i7" class="rowColor">
<td class="colFirst"><code>static <a href="../analysis/Pair.html" title="class in analysis">Pair</a>&lt;java.lang.Integer,java.lang.String&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/StringAnalysis.html#findBestMatch-java.lang.String-java.lang.String-">findBestMatch</a></span>(java.lang.String&nbsp;toAlign,
             java.lang.String&nbsp;template)</code>
<div class="block"><span class="deprecatedLabel">Deprecated.</span>&nbsp;&nbsp;</div>
</td>
</tr>
<tr id="i8" class="altColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/StringAnalysis.html#findHISTag-analysis.AnalysedSequence-">findHISTag</a></span>(<a href="../analysis/AnalysedSequence.html" title="class in analysis">AnalysedSequence</a>&nbsp;toAnalyze)</code>
<div class="block">This method finds the position of the HIS-Tags in the gene.</div>
</td>
</tr>
<tr id="i9" class="rowColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/StringAnalysis.html#findOffset-analysis.AnalysedSequence-">findOffset</a></span>(<a href="../analysis/AnalysedSequence.html" title="class in analysis">AnalysedSequence</a>&nbsp;sequence)</code>
<div class="block">calculates the offset and writes it into the sequence WARNING: does change Offset value!
 
 returns false if begin of sequence was not found Returning false may be an indicator for bad
 sequence, but may also be perfectly fine</div>
</td>
</tr>
<tr id="i10" class="altColor">
<td class="colFirst"><code>static <a href="../analysis/Pair.html" title="class in analysis">Pair</a>&lt;<a href="../analysis/Gene.html" title="class in analysis">Gene</a>,java.lang.Double&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/StringAnalysis.html#findRightGene-analysis.AnalysedSequence-">findRightGene</a></span>(<a href="../analysis/AnalysedSequence.html" title="class in analysis">AnalysedSequence</a>&nbsp;toAnalyze)</code>
<div class="block">Finds the gene that fits best to a given sequence by comparing it to all given genes.</div>
</td>
</tr>
<tr id="i11" class="rowColor">
<td class="colFirst"><code>static <a href="../analysis/Gene.html" title="class in analysis">Gene</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/StringAnalysis.html#findRightGene-analysis.AnalysedSequence-java.util.ArrayList-">findRightGene</a></span>(<a href="../analysis/AnalysedSequence.html" title="class in analysis">AnalysedSequence</a>&nbsp;toAnalyze,
             java.util.ArrayList&lt;<a href="../analysis/Gene.html" title="class in analysis">Gene</a>&gt;&nbsp;listOfGenes)</code>
<div class="block">Finds the gene that fits best to a given sequence by comparing it to all given genes.</div>
</td>
</tr>
<tr id="i12" class="altColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/StringAnalysis.html#findStopcodonPosition-analysis.AnalysedSequence-">findStopcodonPosition</a></span>(<a href="../analysis/AnalysedSequence.html" title="class in analysis">AnalysedSequence</a>&nbsp;toAnalyze)</code>&nbsp;</td>
</tr>
<tr id="i13" class="rowColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/StringAnalysis.html#getLevenshteinIndex-int:A:A-">getLevenshteinIndex</a></span>(int[][]&nbsp;matrix)</code>
<div class="block">gets Levenshtein index out of Levenshtein Matrix</div>
</td>
</tr>
<tr id="i14" class="altColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/StringAnalysis.html#getLevenshteinIndex-java.lang.String-java.lang.String-">getLevenshteinIndex</a></span>(java.lang.String&nbsp;first,
                   java.lang.String&nbsp;second)</code>
<div class="block">calculates Levensthein Matrix of first and second using calculateLevenshteinMatrix(first,
 second) and returns the Levenshtein Index</div>
</td>
</tr>
<tr id="i15" class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/StringAnalysis.html#trimbyLevevenstein-analysis.AnalysedSequence-boolean-">trimbyLevevenstein</a></span>(<a href="../analysis/AnalysedSequence.html" title="class in analysis">AnalysedSequence</a>&nbsp;toAlign,
                  boolean&nbsp;isOffsetExact)</code>
<div class="block"><span class="deprecatedLabel">Deprecated.</span>&nbsp;</div>
</td>
</tr>
<tr id="i16" class="altColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/StringAnalysis.html#trimVector-analysis.AnalysedSequence-">trimVector</a></span>(<a href="../analysis/AnalysedSequence.html" title="class in analysis">AnalysedSequence</a>&nbsp;toAlign)</code>
<div class="block">cuts out the Vector off and writes it into the Left vector of the given sequence Also sets
 Offset @see findOffset()</div>
</td>
</tr>
<tr id="i17" class="rowColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../analysis/StringAnalysis.html#trimVector-analysis.AnalysedSequence-analysis.Gene-">trimVector</a></span>(<a href="../analysis/AnalysedSequence.html" title="class in analysis">AnalysedSequence</a>&nbsp;toAlign,
          <a href="../analysis/Gene.html" title="class in analysis">Gene</a>&nbsp;gene)</code>&nbsp;</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.Object</h3>
<code>equals, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ============ FIELD DETAIL =========== -->
<ul class="blockList">
<li class="blockList"><a name="field.detail">
<!--   -->
</a>
<h3>Field Detail</h3>
<a name="AMINO_ACID_SHORTS">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>AMINO_ACID_SHORTS</h4>
<pre>public static final&nbsp;java.util.Map&lt;java.lang.String,java.lang.String&gt; AMINO_ACID_SHORTS</pre>
<div class="block">A Map with all possible RNA and DNA codons with the matched AminoAcid in short form.</div>
</li>
</ul>
</li>
</ul>
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="StringAnalysis--">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>StringAnalysis</h4>
<pre>public&nbsp;StringAnalysis()</pre>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method.detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="appentStringToLength-java.lang.String-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>appentStringToLength</h4>
<pre>public static&nbsp;java.lang.String&nbsp;appentStringToLength(java.lang.String&nbsp;input,
                                                    int&nbsp;length)</pre>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>input</code> - </dd>
<dd><code>length</code> - </dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
<a name="calculateLevenshteinMatrix-java.lang.String-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>calculateLevenshteinMatrix</h4>
<pre>public static&nbsp;int[][]&nbsp;calculateLevenshteinMatrix(java.lang.String&nbsp;horizontal,
                                                 java.lang.String&nbsp;vertical)</pre>
<div class="block">Calculates the Levensthein Matrix of two Strings. The Matrix gives information about the
 differences of the two Strings and the best way to transform them into another.
 
 for more information: https://en.wikipedia.org/wiki/Levenshtein_distance</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>horizontal</code> - The first String</dd>
<dd><code>vertical</code> - The second String</dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
<a name="checkComplementAndReverse-analysis.AnalysedSequence-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>checkComplementAndReverse</h4>
<pre>public static&nbsp;void&nbsp;checkComplementAndReverse(<a href="../analysis/AnalysedSequence.html" title="class in analysis">AnalysedSequence</a>&nbsp;toAnalyse)
                                      throws <a href="../exceptions/CorruptedSequenceException.html" title="class in exceptions">CorruptedSequenceException</a></pre>
<div class="block">This method checks if the complementary and/or the reverse gene has a better similarity with
 the mutated sequence the the original gene. If this so set reference gene of sequence to the
 best similarity.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>toAnalyse</code> - the sequence to check</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code><a href="../exceptions/CorruptedSequenceException.html" title="class in exceptions">CorruptedSequenceException</a></code> - if complementary cant be build</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../exceptions/CorruptedSequenceException.html" title="class in exceptions"><code>CorruptedSequenceException</code></a></dd>
</dl>
</li>
</ul>
<a name="checkComplementAndReverse-analysis.AnalysedSequence-analysis.Gene-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>checkComplementAndReverse</h4>
<pre>public static&nbsp;void&nbsp;checkComplementAndReverse(<a href="../analysis/AnalysedSequence.html" title="class in analysis">AnalysedSequence</a>&nbsp;toAnalyse,
                                             <a href="../analysis/Gene.html" title="class in analysis">Gene</a>&nbsp;gene)
                                      throws <a href="../exceptions/CorruptedSequenceException.html" title="class in exceptions">CorruptedSequenceException</a></pre>
<div class="block">This method is used when the gene is not the referenced gene in the AnalysedSequence Note: This
 method overrides the old referenced gene with the given gene</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>toAnalyse</code> - the sequence to check</dd>
<dd><code>gene</code> - the gene you reference to.</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code><a href="../exceptions/CorruptedSequenceException.html" title="class in exceptions">CorruptedSequenceException</a></code></dd>
</dl>
</li>
</ul>
<a name="checkSimilarity-analysis.Sequence-analysis.Sequence-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>checkSimilarity</h4>
<pre>public static&nbsp;double&nbsp;checkSimilarity(<a href="../analysis/Sequence.html" title="class in analysis">Sequence</a>&nbsp;first,
                                     <a href="../analysis/Sequence.html" title="class in analysis">Sequence</a>&nbsp;second)</pre>
<div class="block">Compares to sequences and returns their similarity without finding the exact differences.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>first</code> - The first sequence</dd>
<dd><code>second</code> - The second sequence</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>Similarity measure</dd>
</dl>
</li>
</ul>
<a name="checkSimilarity-java.lang.String-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>checkSimilarity</h4>
<pre>public static&nbsp;double&nbsp;checkSimilarity(java.lang.String&nbsp;first,
                                     java.lang.String&nbsp;second)</pre>
<div class="block">Compares to String and returns their similarity without finding the exact differences.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>first</code> - The first String</dd>
<dd><code>second</code> - The second String</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>Similarity measure</dd>
</dl>
</li>
</ul>
<a name="codonsToAminoAcids-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>codonsToAminoAcids</h4>
<pre>public static&nbsp;java.lang.String&nbsp;codonsToAminoAcids(java.lang.String&nbsp;nucleotides)</pre>
<div class="block">Changes the sequence representation from nucleotides to aminoacid (shortform)</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>nucleotides</code> - the sequence presented by nucleotides</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the sequence presented by aminoAcid (shorts)</dd>
</dl>
</li>
</ul>
<a name="findBestMatch-java.lang.String-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>findBestMatch</h4>
<pre>@Deprecated
public static&nbsp;<a href="../analysis/Pair.html" title="class in analysis">Pair</a>&lt;java.lang.Integer,java.lang.String&gt;&nbsp;findBestMatch(java.lang.String&nbsp;toAlign,
                                                                                 java.lang.String&nbsp;template)</pre>
<div class="block"><span class="deprecatedLabel">Deprecated.</span>&nbsp;</div>
<div class="block">Verry slow, do not use</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>toAlign</code> - the String to search in</dd>
<dd><code>template</code> - the String to search for</dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
<a name="findHISTag-analysis.AnalysedSequence-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>findHISTag</h4>
<pre>public static&nbsp;int&nbsp;findHISTag(<a href="../analysis/AnalysedSequence.html" title="class in analysis">AnalysedSequence</a>&nbsp;toAnalyze)</pre>
<div class="block">This method finds the position of the HIS-Tags in the gene.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>toAnalyze</code> - the sequence to analyse</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>returns the startposition of the HIS-Tags. This is relative to the beginning of the
         sequence.</dd>
</dl>
</li>
</ul>
<a name="findRightGene-analysis.AnalysedSequence-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>findRightGene</h4>
<pre>public static&nbsp;<a href="../analysis/Pair.html" title="class in analysis">Pair</a>&lt;<a href="../analysis/Gene.html" title="class in analysis">Gene</a>,java.lang.Double&gt;&nbsp;findRightGene(<a href="../analysis/AnalysedSequence.html" title="class in analysis">AnalysedSequence</a>&nbsp;toAnalyze)</pre>
<div class="block">Finds the gene that fits best to a given sequence by comparing it to all given genes. Known
 genes can be found in the database.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>toAnalyze</code> - Sequence to be analyzed (i.e. for which the fitting gene is to find)</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>The found gene.</dd>
</dl>
</li>
</ul>
<a name="findRightGene-analysis.AnalysedSequence-java.util.ArrayList-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>findRightGene</h4>
<pre>public static&nbsp;<a href="../analysis/Gene.html" title="class in analysis">Gene</a>&nbsp;findRightGene(<a href="../analysis/AnalysedSequence.html" title="class in analysis">AnalysedSequence</a>&nbsp;toAnalyze,
                                 java.util.ArrayList&lt;<a href="../analysis/Gene.html" title="class in analysis">Gene</a>&gt;&nbsp;listOfGenes)</pre>
<div class="block">Finds the gene that fits best to a given sequence by comparing it to all given genes.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>toAnalyze</code> - The sequence, we compare with the list of genes</dd>
<dd><code>listOfGenes</code> - A list of all genes, we want to compare with the sequence</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the gene, that has the best similarity</dd>
</dl>
</li>
</ul>
<a name="findStopcodonPosition-analysis.AnalysedSequence-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>findStopcodonPosition</h4>
<pre>public static&nbsp;int&nbsp;findStopcodonPosition(<a href="../analysis/AnalysedSequence.html" title="class in analysis">AnalysedSequence</a>&nbsp;toAnalyze)</pre>
</li>
</ul>
<a name="getLevenshteinIndex-int:A:A-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getLevenshteinIndex</h4>
<pre>public static&nbsp;int&nbsp;getLevenshteinIndex(int[][]&nbsp;matrix)</pre>
<div class="block">gets Levenshtein index out of Levenshtein Matrix</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>matrix</code> - the Levenshtein Matrix</dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
<a name="getLevenshteinIndex-java.lang.String-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getLevenshteinIndex</h4>
<pre>public static&nbsp;int&nbsp;getLevenshteinIndex(java.lang.String&nbsp;first,
                                      java.lang.String&nbsp;second)</pre>
<div class="block">calculates Levensthein Matrix of first and second using calculateLevenshteinMatrix(first,
 second) and returns the Levenshtein Index</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>first</code> - </dd>
<dd><code>second</code> - </dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
<a name="trimbyLevevenstein-analysis.AnalysedSequence-boolean-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>trimbyLevevenstein</h4>
<pre>@Deprecated
public static&nbsp;java.lang.String&nbsp;trimbyLevevenstein(<a href="../analysis/AnalysedSequence.html" title="class in analysis">AnalysedSequence</a>&nbsp;toAlign,
                                                              boolean&nbsp;isOffsetExact)</pre>
<div class="block"><span class="deprecatedLabel">Deprecated.</span>&nbsp;</div>
<div class="block">same as find Best Match but faster</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>toAlign</code> - </dd>
<dd><code>template</code> - </dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
<a name="trimVector-analysis.AnalysedSequence-analysis.Gene-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>trimVector</h4>
<pre>public static&nbsp;void&nbsp;trimVector(<a href="../analysis/AnalysedSequence.html" title="class in analysis">AnalysedSequence</a>&nbsp;toAlign,
                              <a href="../analysis/Gene.html" title="class in analysis">Gene</a>&nbsp;gene)</pre>
</li>
</ul>
<a name="trimVector-analysis.AnalysedSequence-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>trimVector</h4>
<pre>public static&nbsp;void&nbsp;trimVector(<a href="../analysis/AnalysedSequence.html" title="class in analysis">AnalysedSequence</a>&nbsp;toAlign)</pre>
<div class="block">cuts out the Vector off and writes it into the Left vector of the given sequence Also sets
 Offset @see findOffset()</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>toAlign</code> - </dd>
</dl>
</li>
</ul>
<a name="findOffset-analysis.AnalysedSequence-">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>findOffset</h4>
<pre>public static&nbsp;boolean&nbsp;findOffset(<a href="../analysis/AnalysedSequence.html" title="class in analysis">AnalysedSequence</a>&nbsp;sequence)</pre>
<div class="block">calculates the offset and writes it into the sequence WARNING: does change Offset value!
 
 returns false if begin of sequence was not found Returning false may be an indicator for bad
 sequence, but may also be perfectly fine</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>sequence</code> - </dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/StringAnalysis.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../analysis/Sequence.html" title="class in analysis"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li>Next&nbsp;Class</li>
</ul>
<ul class="navList">
<li><a href="../index.html?analysis/StringAnalysis.html" target="_top">Frames</a></li>
<li><a href="StringAnalysis.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
